app/src/bat_charger.h

```diff
@@ -21,6 +22,193 @@
+#define NUMBER_OF_OBSERVABLES_SOC 1
+
+/**
+ * Extended Kalman Filter (EKF) configuration data
+ *
+ * Data will be initialized in Charger::init_terminal
+ */
+typedef struct
+{
+    /**
+     * number of state values
+     *
+     */
+    int n;
+
+    /**
+     * number of observables
+     *
+     */
+    int m;
+
+    /**
+     * state vector
+     *
+     * [ir0 hk0 SOC0] TODO
+     */
+    float x[NUMBER_OF_STATES_SOC];
+
+    /**
+     * prediction error covariance
+     *
+     */
+    float P[NUMBER_OF_STATES_SOC][NUMBER_OF_STATES_SOC];
+
+    /**
+     * process noise covariance
+     *
+     * uncertainty of current sensor, state equation defined as fx
+     */
+    float Q[NUMBER_OF_STATES_SOC][NUMBER_OF_STATES_SOC];
+
+    /**
+     *  measurement error covariance
+     *
+     * uncertainty of voltage sensor, output equation defined as hx
+     */
+    float R[NUMBER_OF_OBSERVABLES_SOC][NUMBER_OF_OBSERVABLES_SOC];
+
+    /**
+     * Kalman gain; a.k.a. K
+     *
+     */
+    float G[NUMBER_OF_STATES_SOC][NUMBER_OF_OBSERVABLES_SOC];
+
+    /**
+     * Jacobian of process model
+     *
+     */
+    float F[NUMBER_OF_STATES_SOC][NUMBER_OF_STATES_SOC];
+
+    /**
+     * Jacobian of measurement model
+     *
+     */
+    float H[NUMBER_OF_OBSERVABLES_SOC][NUMBER_OF_STATES_SOC];
+
+    /**
+     *  transpose of measurement Jacobian
+     *
+     */
+    float Ht[NUMBER_OF_STATES_SOC][NUMBER_OF_OBSERVABLES_SOC];
+
+    /**
+     * transpose of process Jacobian
+     *
+     */
+    float Ft[NUMBER_OF_STATES_SOC][NUMBER_OF_STATES_SOC];
+
+    /**
+     *  P, post-prediction, pre-update
+     *
+     */
+    float Pp[NUMBER_OF_STATES_SOC][NUMBER_OF_STATES_SOC];
+
+    /**
+     * output of user defined f() state-transition function
+     *
+     */
+    float fx[NUMBER_OF_STATES_SOC];
+
+    /**
+     * output of user defined h() measurement function
+     *
+     */
+    float hx[NUMBER_OF_OBSERVABLES_SOC];
+
+    /**
+     * temporary storage
+     *
+     */
+    float tmp0[NUMBER_OF_STATES_SOC][NUMBER_OF_STATES_SOC];
+    float tmp1[NUMBER_OF_STATES_SOC][NUMBER_OF_OBSERVABLES_SOC];
+    float tmp2[NUMBER_OF_OBSERVABLES_SOC][NUMBER_OF_STATES_SOC];
+    float tmp3[NUMBER_OF_OBSERVABLES_SOC][NUMBER_OF_OBSERVABLES_SOC];
+    float tmp4[NUMBER_OF_OBSERVABLES_SOC][NUMBER_OF_OBSERVABLES_SOC];
+    float tmp5[NUMBER_OF_OBSERVABLES_SOC];
+
+} EkfSoc;
+
+/**
+ * Clamp function to ensure SoC within min max boders.
+ *
+ * @param value SoC value for which to ensure to be within borders min, max
+ * @param min The minimum value 0% the SoC is allowed to have
+ * @param max The maximum value 100% the SoC is allowed to have
+ * @return float Either the handed value or 0%, 100$ when excisting borders.
+ */
+float clamp(float value, float min, float max);
+
+/**
+ * Calculates initial SoC based on handed voltage
+ *
+ * @param battery_voltage_mV
+ * @return float Inital SoC
+ */
+
+float calculate_initial_soc(float battery_voltage_mV);
```

Comment:
Please remove empty line between function and associated comment so that it looks consistent everywhere.